### PR TITLE
Add Supported types of tests to MUnit docs (V2.1)

### PR DESCRIPTION
**Pull request by fermujica**

### CODE REVIEW

**Review by diegostrubolini — APPROVED** [review]

**Review by pentreathm — APPROVED** [review]

File: modules/ROOT/nav.adoc [hunk 1]
```diff
@@ -1,5 +1,6 @@
 .xref:index.adoc[MUnit]
 * xref:index.adoc[Overview]
+ ** xref:supported-types-of-tests.adoc[MUnit Supported Types of Tests]
```
Comment (kbostrom) [inline modules/ROOT/nav.adoc:4]: I see that this is a solitary subtext and wonder if it can be incorporated into the overview container?
You typically want more than one subtopic before you separate information out.

Comment (kbostrom) [reply modules/ROOT/nav.adoc:4]: It would also sound better as Types of Tests Supported by MUnit....

File: modules/ROOT/nav.adoc [hunk 2]
```diff
@@ -1,5 +1,6 @@
...
 
 * xref:configuring-munit.adoc[Configuring MUnit]
```
Comment (kbostrom) [inline modules/ROOT/nav.adoc:6]: HIDDEN TEXT
Edited for capitalization of headings alone; please compare against original, because I cannot edit in hidden text.

xref:build-munit-tutorial.adoc[Building Your First MUnit test]
** xref:build-munit-tutorial-studio.adoc[Using Studio]
** xref:build-munit-tutorial-maven.adoc[Using Maven]
xref:munit-cookbook.adoc[MUnit Examples]
** xref:coverage-config-cookbook.adoc[Configuring Coverage Limits]
** xref:db-utils-cookbook.adoc[Using DB Utilities]
I DON'T KNOW IF WE SHOULD SPELL OUT UTILITIES OR NOT. YOUR DECISION.

** xref:ftp-utils-cookbook.adoc[Using FTP Utilities]
** xref:mock-file-cookbook.adoc[Mocking from File ]
** xref:mock-before-after-foreach-cookbook.adoc[Mocking Message Before and Inside a foreach]
** xref:test-mock-errors-cookbook.adoc[Testing and Mocking Errors]
** xref:environment-config-cookbook.adoc[Testing Environment Configurations]
** xref:utility-classes-dw-cookbook.adoc[Using Utility Classes with a DataWeave Expression from a Test]
** xref:poll-flow-cookbook.adoc[Testing a Flow with a Poll]
** xref:flow-sources-enable-cookbook.adoc[Enabling Flow Sources]

xref:munit-test-concept.adoc[MUnit Test Structure Fundamentals]
** xref:munit-test-reference.adoc[MUnit Test Parameters]
** xref:enable-flow-sources-concept.adoc[Enable Flow Sources]
** xref:before-after-scopes-reference.adoc[Before and After Scopes]
xref:event-processors.adoc[MUnit Processors]
** xref:assertion-event-processor.adoc[Assert]
*** xref:munit-matchers.adoc[Matchers]
THIS IS ANOTHER SOLITARY SUBTOPIC. CAN "ASSERT" BE CHANGED TO "ASSERT MATCHERS" SO THAT WE CAN SKIP THAT LEVEL HEAD AND PROMOTE THE FOLLOWING LEVEL-4 HEADS...BECAUSE WE REALLY DO NOT WANT TO NEST TOPICS BELOW THREE LEVELS?

**** xref:core-matchers-reference.adoc[Core Matchers]
**** xref:string-matchers-reference.adoc[String Matchers]
**** xref:comparable-matchers-reference.adoc[Comparable Matchers]
**** xref:iterable-map-matchers-reference.adoc[Iterable and Map Matchers]
** xref:fail-event-processor.adoc[Fail]
** xref:run-custom-event-processor.adoc[Run Custom Event]
** xref:set-event-processor.adoc[Set Event]
** xref:mock-event-processor.adoc[Mock When]
** xref:verify-event-processor.adoc[Verify]
** xref:spy-event-processor.adoc[Spy Event]
** xref:munit-utils.adoc[Utilities]
*** xref:db-util.adoc[Database]
*** xref:ftp-util.adoc[FTP]

xref:munit-in-studio.adoc[MUnit in Studio]
** xref:munit-update-site.adoc[Update Site]
** xref:munit-scaffold-test-task.adoc[Scaffolding an MUnit Test]
** xref:munit-showing-results.adoc[Showing Results]
** xref:munit-run-configurations.adoc[Run Configurations]
** xref:runtime-patching.adoc[Runtime Patching]
xref:munit-domain-support.adoc[MUnit Support for Domain-Based Applications]
xref:munit-maven-plugin.adoc[MUnit Maven Plug-in]
xref:munit-coverage-report.adoc[Coverage]
** xref:coverage-maven-concept.adoc[Maven Configuration for Coverage]
xref:munit-best-practices.adoc[Best Practices]
I SUGGEST THAT WE AVOID SEPARATE BEST PRACTICES TOPICS IN FAVOR OF INCLUDING THE BEST PRACTICE IN THE INDIVIDUAL TOPICS AND TASKS THEMSELVES. I BELIEVE OUR CREDO IS THAT EVERYTHING WE DOCUMENT IS A BEST PRACTICE, SO THERE SHOULD BE NO SEPARATE LIST, UNLESS THERE IS A BETTER TITLE FOR THIS CONTENT, LIKE "REQUIREMENTS AND RESTRICTIONS" OR SIMILAR.

xref:munit-troubleshooting.adoc[Troubleshooting]

Comment (kbostrom) [reply modules/ROOT/nav.adoc:6]: If you can think of a solution for this limitation, that I cannot edit directly in the hidden text represented by blue bars in these PRs, that would be SWELL! :-)

File: modules/ROOT/pages/supported-types-of-tests.adoc [hunk 3]
```diff
@@ -0,0 +1,41 @@
+= MUnit Supported Types of Tests
```
Comment (kbostrom) [inline modules/ROOT/pages/supported-types-of-tests.adoc:1]:
```suggestion
= Types of Tests Supported by MUnit
```

File: modules/ROOT/pages/supported-types-of-tests.adoc [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+ifndef::env-site,env-github[]
+include::_attributes.adoc[]
+endif::[]
+
+The following are definitions and concepts that MuleSoft recommends to take into account while designing and organizing test cases for your Mule applications. Understanding these definitions and concepts will help you build more robust MUnit tests by covering multiple edge cases, automate your tests through your CI/CD pipeline, and maximize the quality coverage for your Mule application.
```
Comment (kbostrom) [inline modules/ROOT/pages/supported-types-of-tests.adoc:6]: GLOBALLY avoid introducing topics directly by referring to "the following text/section/definitions/concepts...."
Get straight to the point by summarizing the content of the topic set, addressing the audience directly as "you" and telling "you" why they want or need this information, to achieve a specific user goal:
```suggestion
Before you design and organize test cases for your Mule applications, you should understand the types of tests supported by MUnit: unit testing, integration testing, and functional testing. Having this knowledge can help you to build more robust MUnit tests by covering multiple edge cases, to automate your tests through your CI/CD pipeline, and to maximize the quality coverage for your Mule application.
```

File: modules/ROOT/pages/supported-types-of-tests.adoc [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+As a general good practice, MuleSoft recommends following a naming convention for file names and tests so that you can quickly realize what was being tested when a test fails.
```
Comment (kbostrom) [inline modules/ROOT/pages/supported-types-of-tests.adoc:8]:
```suggestion
As a general good practice, be sure to follow a naming convention for file and tests that enables you to quickly identify what was being tested it a test fails.
```

File: modules/ROOT/pages/supported-types-of-tests.adoc [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+When testing a Mule application MUnit allows you to cover three different types of tests:
+
+* Unit tests.
+* Integration tests.
+* Functional test.
```
Comment (kbostrom) [inline modules/ROOT/pages/supported-types-of-tests.adoc:14]: OMIT in favor of introducing the three types right at the beginning, as suggested above.

File: modules/ROOT/pages/supported-types-of-tests.adoc [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+== Unit Testing
+
+Unit testing consists of testing individual units or components of a program. A unit is the smallest testable part of any software. For a Mule application, these units are flows and when doing unit testing for flows, you must isolate them from the rest of the Mule application or external services by mocking these calls to any resource outside the flow you are testing.
```
Comment (kbostrom) [inline modules/ROOT/pages/supported-types-of-tests.adoc:18]:
```suggestion
Unit testing consists of testing the smallest testable part of any software. For a Mule application, this unit is the message flow. When performing unit testing for a message flow, you must isolate it from the rest of the Mule application or external services by mocking the message to a resource outside the flow you are testing.
```

File: modules/ROOT/pages/supported-types-of-tests.adoc [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+
+Creating these tests helps you validate the behavior of key processors and the behavior of a flow as a whole.
```
Comment (kbostrom) [inline modules/ROOT/pages/supported-types-of-tests.adoc:20]:
```suggestion
Performing unit testing helps you to validate the behavior of key processors and of the flow as a whole.
```

File: modules/ROOT/pages/supported-types-of-tests.adoc [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+
+== Integration testing
```
Comment (kbostrom) [inline modules/ROOT/pages/supported-types-of-tests.adoc:22]:
```suggestion
== Integration Testing
```

File: modules/ROOT/pages/supported-types-of-tests.adoc [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+
+When running integration tests you must validate how the units interact with each other or with external systems to expose faults in the interactions. Samples of integration testing in a Mule application are:
+
+* Testing how a service responds (Web Services, Salesforce, ServiceNow, Amazon, etc).
```
Comment (kbostrom) [inline modules/ROOT/pages/supported-types-of-tests.adoc:26]:
```suggestion
* Testing how a service such as a web service, Salesforce, ServiceNow, or Amazon responds
```

Comment (kbostrom) [reply modules/ROOT/pages/supported-types-of-tests.adoc:26]: Should we add what they respond to: "...responds to a message flow"?

File: modules/ROOT/pages/supported-types-of-tests.adoc [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+* Testing a database driver.
```
Comment (kbostrom) [inline modules/ROOT/pages/supported-types-of-tests.adoc:27]:
```suggestion
* Testing a database driver
```

File: modules/ROOT/pages/supported-types-of-tests.adoc [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+* Validating a database response data structure.
```
Comment (kbostrom) [inline modules/ROOT/pages/supported-types-of-tests.adoc:28]:
```suggestion
* Validating a database response data structure
```

File: modules/ROOT/pages/supported-types-of-tests.adoc [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+* Testing in-between flows interactions.
```
Comment (kbostrom) [inline modules/ROOT/pages/supported-types-of-tests.adoc:29]: Does this mean testing the interactions between stages of a flow?
Are the parts of flows called "stages"?
Will the audience wonder if the smallest testable unit is the stage? Prolly not....

File: modules/ROOT/pages/supported-types-of-tests.adoc [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+
+Creating and executing these types of tests requires live versions of the services that your test cases will target.
```
Comment (kbostrom) [inline modules/ROOT/pages/supported-types-of-tests.adoc:31]:
```suggestion
Creating and executing integration tests requires live versions of the services that your test cases can target.
```
"live" vs. what? Aren't unit tests performed against live resources, even if the flows are mocked?

File: modules/ROOT/pages/supported-types-of-tests.adoc [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+
+MuleSoft recommends running the test cases using environment properties and versions of the services dedicated to this end.
```
Comment (kbostrom) [inline modules/ROOT/pages/supported-types-of-tests.adoc:33]: GLOBALLY reword to avoid "MuleSoft recommends" by using imperative voice ("do this") or, as necessary, "you should."

Comment (kbostrom) [reply modules/ROOT/pages/supported-types-of-tests.adoc:33]: Does this sentence refer to the one before it?
Do we mean...
```suggestion
Because integration testing requires live targets, run your test cases using environment properties and versions of the services dedicated to testing.
```

File: modules/ROOT/pages/supported-types-of-tests.adoc [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+
+== Functional testing
```
Comment (kbostrom) [inline modules/ROOT/pages/supported-types-of-tests.adoc:35]:
```suggestion
== Functional Testing
```

File: modules/ROOT/pages/supported-types-of-tests.adoc [hunk 17]
```diff
@@ -0,0 +1,41 @@
...
+
+Functional testing is a wider concept as it means testing the functional requirements of your Mule Application. The goal is to think in your Mule application use case and design test cases that target that functionality.
```
Comment (kbostrom) [inline modules/ROOT/pages/supported-types-of-tests.adoc:37]:
```suggestion
Functional testing involves testing the use cases envisioned for your Mule application as a whole.
```

File: modules/ROOT/pages/supported-types-of-tests.adoc [hunk 18]
```diff
@@ -0,0 +1,41 @@
...
+
+For example, when testing a Mule application that is meant to keep a two-way synchronization between two systems, you can design a test case that creates an entry in each system and verifies the replication in the opposite. If you are testing a batch processing application, you can run a test with a representative set of information and validate the results of the batch. Or if you are testing a service orchestration, you can generate the input and test the result in the other end after the event went through different services.
```
Comment (kbostrom) [inline modules/ROOT/pages/supported-types-of-tests.adoc:39]:
```suggestion
For example, to test a Mule application that is meant to maintain two-way synchronization between two systems, you can design a test case that creates an entry in each system and verifies the replication in the opposite. Or, if you want to test a batch-processing application, you can run a test with a representative set of information and validate the results of the batch. Finally, to test service orchestration, you can generate an event at one end of a flow and test the result at the other end, after the event went through different services.
```

File: modules/ROOT/pages/supported-types-of-tests.adoc [hunk 19]
```diff
@@ -0,0 +1,41 @@
...
+
+MuleSoft recommends running the test cases using environment properties and version of the services dedicated to this end.
```
Comment (kbostrom) [inline modules/ROOT/pages/supported-types-of-tests.adoc:41]:
```suggestion
As with integration testing, run your functional test cases using environment properties and versions of the services dedicated to testing.
```

**Review by kbostrom — APPROVED** [review]
Here you go!
As always, please carefully examine suggested rewordings to ensure that the suggestions do not alter your intended technical meaning. You know best!
Please feel free to Slack me any questions or comments,  Kim

**Comment by fermujica** [comment]
Duplicated content: https://github.com/mulesoft/docs-mule-runtime/pull/915
https://github.com/mulesoft/docs-mule-runtime/pull/933
